Misaligned spans exception when used with spring-data-mongodb-reactive and spring-cloud-stream

## Versions used

Java: 1.8
Spring Boot: 2.4.0-M3
Spring Cloud: 2020.0.0-M4
Spring Cloud Sleuth: 3.0.0-M4

## Description of the error

When Spring Cloud Sleuth is used together with Spring Data MongoDB Reactive and Spring Cloud Stream it throws `Misaligned spans` exceptions.
This is demonstrated by the GitHub project: `git@github.com:magnus-larsson/ml-service-sleuth-issue.git`

This project contains three tests:

1. `createProductDirect` - creates a document in MongoDB using Spring Data MongoDB Reactive
2. `usingStreamWithoutMongoDb` - tests Spring Cloud Stream without using MongoDB
3. `createProductUsingStream` - combines usage of Spring Cloud Stream and Spring Data MongoDB Reactive 

Test 1 and 2 are ok, while test 3 throws an exception:

```
Misalignment: scoped span NoopSpan(57d50a0ff0504383/82431158846274f5) !=  current span LazySpan(57d50a0ff0504383/62a7279d555f7b0e)
java.lang.AssertionError: Misalignment: scoped span NoopSpan(57d50a0ff0504383/82431158846274f5) !=  current span LazySpan(57d50a0ff0504383/62a7279d555f7b0e)
```

## Instructions to reproduce the error

```
git clone git@github.com:magnus-larsson/ml-service-sleuth-issue.git
cd ml-service-sleuth-issue
./gradlew test --info
```

The build will fail and in the log output you can find:

```
> Task :test FAILED

se.magnus.microservices.core.product.MLTests > createProductUsingStream() FAILED
    java.lang.AssertionError: Misalignment: scoped span NoopSpan(1f60cb80f1401bd6/f134e48ecd166acd) !=  current span LazySpan(1f60cb80f1401bd6/4f33d06bb34e1043)
```

To make the failing test run ok, one of the following changes can be made to `build.gradle`:

1. Remove `spring-cloud-starter-sleuth` from the dependencies 
2. Change the Spring Boot version to `2.3.2.RELEASE` and Spring Cloud version to `Hoxton.SR6`

Same problem here with the same description, but with newest versions:
JDK 11
Kotlin 1.4.10
Spring Boot 2.4.0
Spring Cloud 2020.0.0-M5

```
2020-11-25 13:20:29.872 ERROR [app-name_IS_UNDEFINED,abed631d3ccd39ee,f4002f5859cb9edc] 760757 --- [ntLoopGroup-3-4] a.w.r.e.AbstractErrorWebExceptionHandler : [816f22b]  500 Server Error for HTTP PUT "/v1/encodeTypes/code1"

java.lang.AssertionError: Misalignment: scoped span NoopSpan(c48b4623e96c2345/54c16ad747a5cbc9) !=  current span null
	at brave.propagation.ThreadLocalSpan.remove(ThreadLocalSpan.java:164)
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	|_ checkpoint ⇢ Handler cl.santander.txaudit.presentation.EncodeTypeController#update(String, UpdateEncodeTypeDto) [DispatcherHandler]
	|_ checkpoint ⇢ org.springframework.cloud.sleuth.instrument.web.TraceWebFilter [DefaultWebFilterChain]
	|_ checkpoint ⇢ org.springframework.boot.actuate.metrics.web.reactive.server.MetricsWebFilter [DefaultWebFilterChain]
	|_ checkpoint ⇢ HTTP PUT "/v1/encodeTypes/code1" [ExceptionHandlingWebHandler]
Stack trace:
		at brave.propagation.ThreadLocalSpan.remove(ThreadLocalSpan.java:164)
		at brave.mongodb.TraceMongoCommandListener.commandSucceeded(TraceMongoCommandListener.java:103)
		at com.mongodb.internal.connection.ProtocolHelper.sendCommandSucceededEvent(ProtocolHelper.java:289)
		at com.mongodb.internal.connection.LoggingCommandEventSender.sendSucceededEvent(LoggingCommandEventSender.java:142)
```

We're also experiencing this problem, it is preventing us from upgrading to spring boot 2.4.0.

We're using:

JDK 14
Kotlin 1.4.21
Spring Boot 2.4.0
Spring Cloud 2020.0.0-M6
Gcp dependencies: 1.2.6.RELEASE
spring-cloud-starter-sleuth from Spring Cloud 2020.0.0-M6
spring-cloud-gcp-starter-trace from Spring Cloud 2020.0.0-M6
spring-boot-starter-data-mongodb-reactive

```
Caused by: java.lang.AssertionError: Misalignment: scoped span NoopSpan(25a281266543b1d3/cb424d6e4c2b7966) !=  current span null
	at brave.propagation.ThreadLocalSpan.remove(ThreadLocalSpan.java:164)
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Assembly trace from producer [reactor.core.publisher.MonoLift] :
	reactor.core.publisher.Mono.from(Mono.java:488)
	org.springframework.data.mongodb.core.ReactiveMongoTemplate.lambda$executeFindOneInternal$89(ReactiveMongoTemplate.java:2762)
```

Stack trace:

```
Stack trace:
		at brave.propagation.ThreadLocalSpan.remove(ThreadLocalSpan.java:164)
		at brave.mongodb.TraceMongoCommandListener.commandSucceeded(TraceMongoCommandListener.java:103)
		at com.mongodb.internal.connection.ProtocolHelper.sendCommandSucceededEvent(ProtocolHelper.java:289)
		at com.mongodb.internal.connection.LoggingCommandEventSender.sendSucceededEvent(LoggingCommandEventSender.java:142)
		at com.mongodb.internal.connection.InternalStreamConnection$2$1.onResult(InternalStreamConnection.java:460)
		at com.mongodb.internal.connection.InternalStreamConnection$2$1.onResult(InternalStreamConnection.java:440)
		at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback$MessageCallback.onResult(InternalStreamConnection.java:745)
		at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback$MessageCallback.onResult(InternalStreamConnection.java:712)
		at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:582)
		at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:579)
		at com.mongodb.connection.netty.NettyStream.readAsync(NettyStream.java:255)
		at com.mongodb.connection.netty.NettyStream.readAsync(NettyStream.java:214)
		at com.mongodb.internal.connection.InternalStreamConnection.readAsync(InternalStreamConnection.java:579)
		at com.mongodb.internal.connection.InternalStreamConnection.access$1100(InternalStreamConnection.java:78)
		at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback.onResult(InternalStreamConnection.java:702)
		at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback.onResult(InternalStreamConnection.java:687)
		at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:582)
		at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:579)
		at com.mongodb.connection.netty.NettyStream.readAsync(NettyStream.java:255)
		at com.mongodb.connection.netty.NettyStream.readAsync(NettyStream.java:214)
		at com.mongodb.connection.netty.NettyStream.handleReadResponse(NettyStream.java:285)
		at com.mongodb.connection.netty.NettyStream.access$800(NettyStream.java:69)
		at com.mongodb.connection.netty.NettyStream$InboundBufferHandler.channelRead0(NettyStream.java:344)
		at com.mongodb.connection.netty.NettyStream$InboundBufferHandler.channelRead0(NettyStream.java:341)
		at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		at java.base/java.lang.Thread.run(Thread.java:832)
```

--- a/build.gradle
+++ b/build.gradle
@@ -1,6 +1,5 @@
 plugins {
-    id 'org.springframework.boot' version '2.4.0-M3'
-//  id 'org.springframework.boot' version '2.3.2.RELEASE'
+    id 'org.springframework.boot' version '2.4.0-RC1'
     id 'io.spring.dependency-management' version '1.0.10.RELEASE'
     id 'java'
 }
@@ -16,8 +15,7 @@ repositories {
 
 ext {
     mapstructVersion = "1.3.1.Final"
-    springCloudVersion = "2020.0.0-M4"
-//  springCloudVersion = "Hoxton.SR6"
+    springCloudVersion = "2020.0.0-M6"
 }
 
 dependencies {

